### PR TITLE
Image Upload and profile edit integration with backend.

### PR DESCRIPTION
**Pull request by tejasj1718**
### What this PR does (required):
- Completes the integration with the front end and the back end.
- Added additional routes added to ease integration.
- Live update of the profile photo.

-NEW Upload Photo Dialog

### Screenshots / Videos (front-end only):
- Old Upload profile photo
PR4
New Upload Profile Photo
PRImageUpload
- Edit your profile.
PR2

### Any information needed to test this feature (required):
- Clear your MongoDB documents in case you have an older version of the schema.
- Create a new user and login.
- Click on the avatar to upload a profile photo.
- Click on the extra setting icon and click on edit profile.
- Edit the profile settings and watch the username change on the dashboard. Use the /auth/user Get request to check the new user details.
- Change the password and try to login with the new password.

### Any issues with the current functionality (optional):
- ~~The profile photo dialog is not well designed.~~

### CODE REVIEW

File: server/models/Image.js
Comment (ugur-akin) [inline server/models/Image.js:25]: Look into mongoose timestamps for this ([docs](https://mongoosejs.com/docs/guide.html#timestamps)).

File: server/controllers/helper.js
Comment (ugur-akin) [inline server/controllers/helper.js:21]: Similar comments about single-responsibility here. Ideally this function is only concerned about finding out whether the user exists or not - setting status codes and/or sending responses shouldn't be its responsibility.

File: server/models/User.js
Comment (ugur-akin) [inline server/models/User.js:27]: 👍🏼

File: client/src/pages/EditProfile/EditProfile.tsx
Comment (ugur-akin) [inline client/src/pages/EditProfile/EditProfile.tsx:71]: We won't need to do this when we implement `ProtectedRoute`s - can you either remove it or leave a comment to remind about this?

File: client/src/pages/EditProfile/EditProfileForm/EditProfileForm.tsx
Comment (ugur-akin) [inline client/src/pages/EditProfile/EditProfileForm/EditProfileForm.tsx:18]: Refactoring this into a quick interface would improve the readability of this code-block significantly 👍🏼

File: client/src/pages/EditProfile/EditProfileForm/EditProfileForm.tsx
Comment (ugur-akin) [inline client/src/pages/EditProfile/EditProfileForm/EditProfileForm.tsx:43]: I generally recommend defining these objects before render/return (e.g. outside of the JSX) to promote clarity and readability 👍🏼 

Material UI is particularly good at this - check out how clear the UI of [`Button`](https://github.com/mui-org/material-ui/blob/c545ccab7edfdf4a44d4ec2f4bf10ebc7fd00259/packages/material-ui/src/Button/Button.js#L294) and [`TextField`](https://github.com/mui-org/material-ui/blob/c545ccab7edfdf4a44d4ec2f4bf10ebc7fd00259/packages/material-ui/src/TextField/TextField.js#L160) components are despite their complex implementations!

File: client/src/pages/EditProfile/ChangePasswordForm/ChangePasswordForm.tsx
Comment (ugur-akin) [inline client/src/pages/EditProfile/ChangePasswordForm/ChangePasswordForm.tsx:48]: Similar concerns here 👍🏼

File: client/src/helpers/APICalls/getProfilePhoto.ts
Comment (ugur-akin) [inline client/src/helpers/APICalls/getProfilePhoto.ts:11]: A content-type header isn't needed here (see [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type) for details)

File: client/src/components/ChatSideBanner/ChatSideBanner.tsx
Comment (ugur-akin) [inline client/src/components/ChatSideBanner/ChatSideBanner.tsx:57]: It would be better to implement this menu/dialog on the navbar avatar rather than the ChatSidebar (our project doesn't have a sidebar in mocks, correct?).

**Review by ugur-akin — CHANGES_REQUESTED** [review]
Whew 💦 Lot's a changes here - this is both nice to see because of the good progress 🎉  and scary because it's difficult to go over as a reviewer😅In the future, please keep your PRs smaller by only tackling a single issue/feature and committing only the relevant changes. 

Left a few remarks here and there, take a look and feel free to initiate any further discussions! One additional thing I want to insist on is that we adjust the styling of the upload dialog to match the designs in the mocks closer before we merge this feature 👍🏼 

Really like your PR title, it would also be really helpful to [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) on GitHub as well (this is different than ZenHub's issue/PR connection) 👍🏼 

Keep up the good work!

File: client/src/components/ImageUpload/ImageUpload.tsx
Comment (tejasj1718) [reply client/src/components/ImageUpload/ImageUpload.tsx:41]: This issue is caused due to ` <Input inputProps={state.getInputProps()} />` giving errors with the typing so that is why I had to do that. Type 'DropzoneInputProps' is not assignable to type 'InputBaseComponentProps'. The type set by react dropzone does not match the type required by MUI, I have tried to fix this and came up with this solution.

File: client/src/components/ChatSideBanner/ChatSideBanner.tsx
Comment (tejasj1718) [reply client/src/components/ChatSideBanner/ChatSideBanner.tsx:57]: Yes, this feature was implemented before the navbar was made. Once this PR is approved I will add the components to the new design.

File: client/src/helpers/APICalls/changePassword.ts
Comment (tejasj1718) [reply client/src/helpers/APICalls/changePassword.ts:23]: Since my PR requires me to edit the profile I added this in as well, I hope it's ok to leave it in for now for future use.

File: server/controllers/image.js
Comment (tejasj1718) [reply server/controllers/image.js:45]: That was the intended behaviour as the Image model accepts a null user id. So we have the option to upload without needing a protected route.

File: client/src/pages/EditProfile/EditProfile.tsx
Comment (tejasj1718) [reply client/src/pages/EditProfile/EditProfile.tsx:71]: I Will leave a comment, will add it in during merge into main as the protected route is already implemented.

File: server/controllers/image.js
Comment (tejasj1718) [reply server/controllers/image.js:39]: will refactor this and throw errors only in the controllers.

File: server/controllers/image.js
Comment (tejasj1718) [reply server/controllers/image.js:22]: I have decided to not let the helper function accept req and res as per your earlier comment which I agree, and I will handle the errors in the controllers which should take advantage of asyncHandler

File: client/src/pages/EditProfile/EditProfileForm/EditProfileForm.tsx
Comment (tejasj1718) [reply client/src/pages/EditProfile/EditProfileForm/EditProfileForm.tsx:43]: I think I understand what you mean but can you clarify what you mean by how clear the UI is?

File: server/controllers/helper.js
Comment (tejasj1718) [reply server/controllers/helper.js:21]: will try to create a workaround that you might approve.

File: client/src/components/ImageUpload/ImageUpload.tsx
Comment (tejasj1718) [reply client/src/components/ImageUpload/ImageUpload.tsx]: image

File: client/src/components/ImageUpload/useStyles.ts
Comment (tejasj1718) [reply client/src/components/ImageUpload/useStyles.ts:19]: will look into this

File: server/controllers/auth.js
Comment (ugur-akin) [inline server/controllers/auth.js]: Nice use of the error handler! 💯  I would say [`422 Unprocessable Entity`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422) fits this better than `400` (for what it's worth, it's also what GitHub uses when the request for its API fails validation).

File: server/controllers/image.js
Comment (ugur-akin) [inline server/controllers/image.js:62]:
```suggestion
  const user = await User.findById(req.user.id);
```
`const` doesn't mean the variable is immutable, only that its not re-assignable 👍🏼 Unless, you are supposed to have re-assingment of `user = ...` somewhere, we should be using `const`. (you can give [this article](https://dev.to/savagepixie/mutability-and-reassignability-in-javascript-10on) a read if you're unfamiliar with the difference)

File: server/controllers/image.js
Comment (ugur-akin) [inline server/controllers/image.js:77]: 💯

**Review by ugur-akin — CHANGES_REQUESTED** [review]
Well done with this! 💯 We can merge this in after a couple more minor fixes and resolving merge conflicts - give me a (or your team lead a ping when it's ready). 

Good job on fixing the type errors, it's the sort of stuff that can be annoying about TypeScript but it's usually for a good reason!

You were spot on with the changes about having a more readable UI/JSS, but to answer your question - it's much easier to follow what a component renders if we keep the the JSS we return from it clean of logic/JS. 
Best practices in this regard usually involve avoiding inline definitions of functions/objects passed to props, abstracting conditionals (or other logic) outside of the return statement and breaking down the UI into smaller components as necessary 👍🏼

**Review by moffatethan — APPROVED** [review]
Good work!

**Comment by moffatethan** [comment]
Please fix all merge conflicts before merging!

**Pull request merged**
